### PR TITLE
inorder, preorder, postorder, bfs

### PR DESCRIPTION
**Pull request by albertywu**
solution to tree traversal problem

### CODE REVIEW

File: coding-harder/TreeTraversal.js [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+  }
+
+  size() {
+    function _size(node) {
+      return node === null ? 0 : 1 + _size(node.left) + _size(node.right)
+    }
+    return _size(this.root)
+  }
+
+  inorder() {
```
Comment (bcherny) [inline coding-harder/TreeTraversal.js]: Should this take a function?

File: coding-harder/TreeTraversal.js [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+      if (node === null) return
+      _inorder(node.left)
+      queue.enqueue(node)
+      _inorder(node.right)
+    }
+    _inorder(this.root)
+    return queue
+  }
+
+  preorder() {
```
Comment (bcherny) [inline coding-harder/TreeTraversal.js]: Same here

File: coding-harder/TreeTraversal.js [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+      if (node === null) return
+      queue.enqueue(node)
+      _preorder(node.left)
+      _preorder(node.right)
+    }
+    _preorder(this.root)
+    return queue
+  }
+
+  postorder() {
```
Comment (bcherny) [inline coding-harder/TreeTraversal.js]: And here

File: coding-harder/TreeTraversal.js [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+      if (node === null) return
+      _postorder(node.left)
+      _postorder(node.right)
+      queue.enqueue(node)
+    }
+    _postorder(this.root)
+    return queue
+  }
+
+  bfs() {
```
Comment (bcherny) [inline coding-harder/TreeTraversal.js]: And here

File: coding-harder/TreeTraversal.js [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+  isEmpty() { return this.data.length === 0 }
+
+  [Symbol.iterator]() {
+    let idx = this.data.length - 1
+    return {
+      next: () => idx > -1
+        ? { value: this.data[idx--], done: false }
+        : { done: true }
+    }
+  }
```
Comment (bcherny) [inline coding-harder/TreeTraversal.js]: This is really nice! You might consider making this a generator and using `yield` to reduce some of this boilerplate!

File: coding-harder/TreeTraversal.js [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+    { key: 'a', value: 42 },
+    { key: 'd', value: 100 },
+    { key: 'i', value: 100 },
+    { key: 'c', value: 100 },
+    { key: 'e', value: 100 },
+    { key: 'h', value: 100 }
+  )
+
+  let nodes = Array
+    .from(bst.bfs())
```
Comment (bcherny) [inline coding-harder/TreeTraversal.js]: Neat!!

**Review by bcherny — CHANGES_REQUESTED** [review]
Looks great overall - some minor comments.

**Comment by albertywu** [comment]
done!

**Review by bcherny — APPROVED** [review]
Looking great! Thanks @albertywu!

**Pull request merged**